### DISCUSSION TITLE
Add a NameOf operator with two arguments

### DISCUSSION THREAD

**Discussion by kingtu** [dotnet/csharplang#4992] [Language Ideas] [answered]
//const string FullProductName = $"{Platform} - Language: {Language} Version: {Version}";
string s=NameOf(FullProductName ,(x) x.ToUpper) // result is FULLPRODUCTNAME
The second argument is a lambda expression。

**Comment by YairHalberstadt** [comment, marked as answer]
Why not just do `nameof(FullProductName).ToUpper()`? That's shorter and easier to understand, and legal today.

**Reply by spydacarnage** [reply]
Also, `nameof()` creates a compile-time constant, which could no longer be the case if you need to apply a lambda function to produce the name.

**Reply by kingtu** [reply]
switch (name)
        {
            case “n1”:
                break;
           case NameOf(FullProductName ,(x) x.ToUpper) :
                break;
            default:
                break;
        }

Want to make the above syntax legal.

**Reply by YairHalberstadt** [reply]
Yes, but why?

**Reply by kingtu** [reply]
It is mainly used in switch statements。
A case in a switch statement cannot use a variable

**Reply by spydacarnage** [reply]
It can only use compile-time constants, yes.  But by adding a lambda function into `nameof()`, you make it no longer compile-time, because functions like that can only be run at run-time.

**Reply by kingtu** [reply]
What is the way to achieve this function?

**Reply by spydacarnage** [reply]
It isn't possible.
You either have to modify the input value to match your compile-time constant, or if you have to compare to a runtime value, you'll have to use `if else`.

**Reply by ufcpp** [reply]
ToUpper depends on CurrentCulture, and thus might not be suitable for compile-time constant.

https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1TATgAoAiASzIDoAVAewFUAHZgUzBIEouBuPApjoALMGwCGcClADmIsZOoBhCGDFRgtURLjLV64CoA2wVWwAEAXnNQ2Ad3OFqAcSP0ARuKMUAXuOAU9FB6JmYAklAAZvTkwGAAtLQASmR8gqSUNAws7Jw8/LhAA===

**Reply by myangelkamikaze** [reply]
For this case you could change it into a pattern instead. 

```cs
switch (name)
{
   case “n1”:
      break;
   case string s when s == nameof(FullProductName).ToUpper():
      break;
   default:
      break;
}
```
